src/RedisIPCMessenger.ts

```diff
@@ -0,0 +1,119 @@
+import Redis from 'ioredis';
+import IPCMessenger, {
+  Room,
+  Instance,
+  MessageCallback,
+  MessageTypes,
+  Message,
+} from './IPCMessenger';
+
+type ConstructorParams = {
+  room: Room;
+  instance: Instance;
+  callback: MessageCallback,
+  redis?: Redis.RedisOptions,
+  expireTime?: number,
+  refreshTime?: number,
+}
+
+class RedisIPCMessenger implements IPCMessenger {
+  private readonly room: Room;
+  private readonly instance: Instance;
+  private readonly callback: MessageCallback;
+  private readonly publisher: Redis.Redis;
+  private readonly subscriber: Redis.Redis;
+  private readonly expireTime: number;
+  private readonly refreshTime: number;
+  private hasJoined: boolean;
+
+  constructor(params: ConstructorParams) {
+    this.validateInput(params);
+
+    this.room = params.room;
+    this.instance = params.instance;
+    this.callback = params.callback;
+
+    this.publisher = new Redis(params.redis);
+    this.subscriber = new Redis(params.redis);
+    this.expireTime = params.expireTime ?? 30000;
+    this.refreshTime = params.refreshTime ?? 10000;
+    this.hasJoined = false;
+  }
+
+  async join(): Promise<void> {
+    if (this.hasJoined) {
+      throw new Error('Can\'t join twice!');
+    }
+
+    const refreshKeyLoop = async () => {
+      await this.publisher.set(`${this.room}:${this.instance}`, '', 'EX', this.expireTime / 1000);
+      setTimeout(refreshKeyLoop, this.refreshTime);
+    };
+    await refreshKeyLoop();
+
+    await this.subscriber.config('SET', 'notify-keyspace-events', 'Kx');
+    await this.subscriber.psubscribe(`__keyspace@0__:${this.room}:*`);
+    this.subscriber.on('pmessage', (pattern, key) => {
```

Comment:
Why don't we extract the callback on the class level, something like `onPMessage`?

Reply:
Done, picked a more descriptive name though - onExpiredKeyMessage/onPubSubMessage